### ISSUE TITLE
Website development

### ISSUE THREAD

**Issue by dannycolin** [alexmorley/meta-open-database#3]
This issue is to talk about the technical aspect of the database development 

- Tools
- Hosting
- Functionalities (e.g. API)
- Licensing (see #5)

**Comment by alexmorley** [comment]
Can we split these up a bit?
Ideas for things we can consider relatively separately.

- UI/Front-end
- REST API Endpoints
- Database tooling (I've given this some thought and think MongoDB is going to be the most flexible/lowest barrier for this. But we can create an issue if someone has a different preference).
- Hosting, happy to discuss this now but might end up depending on what resources we have available. I have enough credits on Azure, GCP and digital ocean for development.
- Licensing. For sure this could use it's own issue and is worth thinking hard about.

**Comment by dannycolin** [comment]
> Can we split these up a bit?

Yes. Of Course. I've opened this issue exactly to see what we need and how we should divide it in sub-issues/tasks.  

> Database tooling (I've given this some thought and think MongoDB is going to be the most flexible/lowest barrier for this

I've never used MongoDB personally but I've heard good things about it. I'll just point out one (temporary) caveat. they've recently relicensed it under a new license called Server Side Public License that haven't been OSI/FSF approved yet. [There's a FAQ page](https://www.mongodb.com/licensing/server-side-public-license/faq) if you need more information.

**Comment by jcolomb** [comment]
about database tooling, we probably need different tools for gathering information (low tech barrier, i.e. probably google spreadsheets/survey tools) and getting the database in state of the art tools (there I have no idea).

In terms of website, I think the discussion jekyll versus hugo (i.e. Rblowdown ?) should happen soon? 

In addition, we should think about making it datproject ready (i.e. no https dependencies).

**Comment by dannycolin** [comment]
> about database tooling, we probably need different tools for gathering information (low tech barrier, i.e. probably google spreadsheets/survey tools)

Ideally, there'll be a way to submit/edit the data live in the website interface. In the meantime, we could use plaintext file with JSON, YAML or any other data serialization language that we can easily programatically import in a more robust database later. The advantages are that we could host the file here on GitHub and version them and anyone can fork them easily. 

> In terms of website, I think the discussion jekyll versus hugo (i.e. Rblowdown ?) should happen soon?

We'll probably need something more complex than a static site generator especially if we want to implement a search by criteria functionality. But if we think a static site is enough for now then IMO Hugo offers a lot more functionality and flexibility than Jekyll.

> In addition, we should think about making it datproject ready

Oh I like the idea of supporting dat protocol :).

**Comment by alexmorley** [comment]
> 
> > about database tooling, we probably need different tools for gathering information (low tech barrier, i.e. probably google spreadsheets/survey tools)
> 
> Ideally, there'll be a way to submit/edit the data live in the website interface. In the meantime, we could use plaintext file with JSON, YAML or any other data serialization language that we can easily programatically import in a more robust database later. The advantages are that we could host the file here on GitHub and version them and anyone can fork them easily.

Yeah that's the way I was thinking about this. JSON strikes me as the easiest way to go with this for now as it's easy-ish to update the schema as we go. (Also this makes it trivial to use with dat).


> We'll probably need something more complex than a static site generator especially if we want to implement a search by criteria functionality. But if we think a static site is enough for now then IMO Hugo offers a lot more functionality and flexibility than Jekyll.
> 
> > In addition, we should think about making it datproject ready
> 
> Oh I like the idea of supporting dat protocol :).

Yes we'll need to use a dynamic website to support user submissions and API requests. The way I usually do this is to use nodejs (with the Express framework) for the sole reason that it means we can use the same language on the front- and back-end (i.e. it makes things simple/easy).

**Comment by dannycolin** [comment]
> The way I usually do this is to use nodejs (with the Express framework)

Well I'm not a fan of nodejs' ecosystem. But you're right it's easier to use the same language. Plus, we already both know JS so that's an other advantage to use it. ;)

So there's still one thing about the devstack we should look at. 

> [MongoDB] recently relicensed [their source code] under a new license called Server Side Public License that hasn't been OSI/FSF approved yet.

@alexmorley what's your opinion on that?

**Comment by alexmorley** [comment]
> Plus, we already both know JS so that's an other advantage to use it. ;)

This seems like a pretty major plus! :+1: 

>  >  [MongoDB] recently relicensed [their source code] under a new license called Server Side Public License that hasn't been OSI/FSF approved yet.
> 
> @alexmorley what's your opinion on that?

I was planning on maintaining a copy of the database as (a collection of) json files. So my lazy answer would be that it's doesn't matter for now and as long as we make sure we don't entangle anything that is mongo-specific. But then again I haven't actually read the license so I should probably go ahead and do that now...

**Comment by dannycolin** [comment]
@alexmorley https://www.zdnet.com/article/mongodb-open-source-server-side-public-license-rejected/ Looking bad for MongoDB 😕

**Comment by dannycolin** [comment]
Follow up on MongoDB :
Debian [1], Red Hat [2] and Arch Linux [3] distributions have decided to remove it from their repository.

- [1] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=916107#5
- [2] https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8-beta/html/8.0_beta_release_notes/new-features#web_servers_databases_dynamic_languages_2
- [3] https://lists.archlinux.org/pipermail/arch-dev-public/2019-January/029430.html

**Comment by alexmorley** [comment]
Yeah it's not ideal. Although I do sympathize with them a bit. The API I'm using at the moment predates this license anyway... and I'm kind of hoping that enough people use (even Amazon do in DocumentDB) it that someone is going to maintain something compatible with a OSI approved license. Is that wishful thinking? If so are there similar JSON-based noSQL alternatives that you have experience with?

**Comment by dannycolin** [comment]
> I'm kind of hoping that enough people use (even Amazon do in DocumentDB) it that someone is going to maintain something compatible with a OSI approved license.

I hope too! I think will probably see an announcement from MongoDB before the end of february.

> If so are there similar JSON-based noSQL alternatives that you have experience with?

Well, there's not a lot of alternative out there that have a good userbase. The only that I can think of is CouchDB. I never use it personally but they seems to have a nice REST API.
